### PR TITLE
barista(component-pages): Generate json output for all component pages.

### LINKED ISSUE #133: Barista: generate JSON output for md-files in components/core
Currently we only generate component json files for components that are a direct child directory of the components directory (and all documentation pages from root and the documentation dir). But the "Indicator" and the "Viewport resizer" components/directives/services are still missing.

Recursively look for all `README.md` files and `barista.json` files to not miss anything.

### PR DESCRIPTION
**Pull request by NinaKammerlander**
## <strong>Pull Request</strong>

fixes #133

<hr>
Hi, thank you for contributing to the Dynatrace Angular Component library (PR).

To ensure a fast process and merging of your PR please make sure it fulfills the
coding standards and contribution guidelines.

- A feature proposal has been provided, discussed and approved first.
- There is a meaningful description of the issue in GitHub (Screenshots are
  often helpful).
- If the PR introduces breaking-changes or deprecations it matches the following
  guidelines.
  - The commit message follows our commit guidelines.
  - Tests for the changes have been added (for bug fixes / features).
  - Docs have been added / updated (for bug fixes / features).

Please fill out the checkbox below: <br>

#### Type of PR

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or change that would cause existing functionality to
      not work as expected)
- [ ] Documentation update (changes to documentation)
- [ ] Other (if none of the above apply)

#### Checklist

- [ ] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tools/barista/src/builder/components.ts [hunk 1]
```diff
@@ -73,22 +73,39 @@ function setMetadataDefaults(baristaMetadata: any): BaSinglePageMeta {
   return metadataWithDefaults;
 }
 
+function getPaths(path: string, fileList: string[]): string[] {
+  let files = readdirSync(path);
+
+  files.forEach(function(file: string): void {
```
Comment (thomaspink) [inline tools/barista/src/builder/components.ts]: Pls use the forof loop

File: tools/barista/src/builder/components.ts [hunk 2]
```diff
@@ -73,22 +73,39 @@ function setMetadataDefaults(baristaMetadata: any): BaSinglePageMeta {
...
+    const filePath = join(path, file);
+    if (lstatSync(filePath).isDirectory()) {
+      // Only grab those dirs that include a README.md and a barista.json
+      if (
+        existsSync(join(filePath, 'README.md')) &&
+        existsSync(join(filePath, 'barista.json'))
+      ) {
+        fileList.push(filePath);
+      }
+      getPaths(filePath, fileList);
```
Comment (thomaspink) [inline tools/barista/src/builder/components.ts]: i guess the paths returned here should also be pushed

File: tools/barista/src/builder/components.ts [hunk 3]
```diff
@@ -73,22 +73,39 @@ function setMetadataDefaults(baristaMetadata: any): BaSinglePageMeta {
...
-  // Only grab those dirs that include a README.md and a barista.json
-  const readmeDirs = paths.filter(
-    dir =>
-      existsSync(join(dir, 'README.md')) &&
-      existsSync(join(dir, 'barista.json')),
-  );
+  let readmeDirs: string[] = [];
+
+  if (componentsPaths) {
+    componentsPaths.forEach(path => {
```
Comment (thomaspink) [inline tools/barista/src/builder/components.ts]: Pls use forof loop here

File: tools/barista/src/builder/components.ts [hunk 4]
```diff
@@ -73,22 +73,39 @@ function setMetadataDefaults(baristaMetadata: any): BaSinglePageMeta {
   return metadataWithDefaults;
 }
 
+function getPaths(path: string, fileList: string[]): string[] {
```
Comment (thomaspink) [inline tools/barista/src/builder/components.ts]: This function should also check for the path provided, not only its sub directories

File: tools/barista/src/builder/components.ts (hunk 2 above)
Comment (tomheller) [reply tools/barista/src/builder/components.ts]: this should do the trick. 
```suggestion
      fileList.push(...getPaths(filePath, fileList));
```

File: tools/barista/src/builder/components.ts [hunk 5]
```diff
@@ -73,22 +73,39 @@ function setMetadataDefaults(baristaMetadata: any): BaSinglePageMeta {
...
-  const readmeDirs = paths.filter(
-    dir =>
-      existsSync(join(dir, 'README.md')) &&
-      existsSync(join(dir, 'barista.json')),
-  );
+  let readmeDirs: string[] = [];
+
+  if (componentsPaths) {
+    componentsPaths.forEach(path => {
+      readmeDirs = getPaths(path, readmeDirs);
```
Comment (tomheller) [inline tools/barista/src/builder/components.ts]: You probably want to join all paths here so either push to the readMeDirs, or spread them together: 
```suggestion
      readmeDirs.push(...getPaths(path, readmeDirs));
```

or 
```suggestion
      readmeDirs = [...readmeDirs, ...getPaths(path, readmeDirs));
```

**Comment by thomaspink** [comment]
LGTM

File: tools/barista/src/builder/components.ts [hunk 6]
```diff
@@ -73,22 +73,41 @@ function setMetadataDefaults(baristaMetadata: any): BaSinglePageMeta {
...
+  ) {
+    fileList.push(path);
+  }
+
+  let files = readdirSync(path);
+
+  for (const file of files) {
+    const filePath = join(path, file);
+    if (lstatSync(filePath).isDirectory()) {
+      // Only grab those dirs that include a README.md and a barista.json
```
Comment (lara-aigmueller) [inline tools/barista/src/builder/components.ts]: Move comment to line 78 ;-)

**Pull request merged**